tests/kernel/smp/src/main.c

```diff
@@ -72,20 +111,209 @@ void test_main(void)
 	}
 
 	for (i = 0; i < 10; i++) {
-		/* Wait slightly longer than the other thread so our
-		 * count will always be lower
-		 */
 		k_busy_wait(DELAY_US + (DELAY_US / 8));
 
-		if (t2_count <= i) {
-			ok = 0;
-			break;
-		}
+		zassert_true(t2_count > i, "Another thread isn't"
+			     " scheduled to run on other core");
+	}
+	k_thread_abort(tid);
+}
+
+static void smp_thread_entry(void *p1, void *p2, void *p3)
+{
+	int tnum = (int)p1;
+
+	k_busy_wait(DELAY_US);
+
+	tdata[tnum].executed = 1;
+}
+
+static void spawn_threads(int thread_num, int prio, int start_delay)
+{
+	int i;
+
+	for (i = 0; i < thread_num; i++) {
+		tdata[i].executed = 0;
+		tdata[i].tid = k_thread_create(&tthread[i], tstack[i],
+					       STACK_SIZE, smp_thread_entry,
+					       (void *)i, NULL, NULL,
+					       prio, 0, start_delay);
+	}
+}
+
+/**
+ * @brief Verify round robin scheduling with one core busy
+ *
+ * @details The test verifies scenario in which parent
+ * thread which doesn't want itself to be preempted
+ * spawns threads of equal priority to complete a task
+ *
+ */
+static void test_round_robin_other_cores(void)
+{
+	int i;
+
+	/* Set time slice */
+	k_sched_time_slice_set(TIME_SLICE_MS, 0);
+
+	spawn_threads(THREADS_NUM,
+		      K_LOWEST_APPLICATION_THREAD_PRIO, 0);
+
+	/* Round robin scheduling should happen
+	 * on all cores except the current one
+	 */
+	k_busy_wait((DELAY_US * THREADS_NUM) /
+		    (CONFIG_MP_NUM_CPUS - 1));
+
+	for (i = 0; i < THREADS_NUM; i++) {
+		k_thread_abort(tdata[i].tid);
+	}
+
+	/* Reset the time slicing */
+	k_sched_time_slice_set(0, 0);
+
+	/* Test point: All THREADS_NUM spawned threads must have
+	 * executed in another core
+	 */
+	for (i = 0; i < THREADS_NUM; i++) {
+		zassert_true(tdata[i].executed == 1, "Threads didn't get"
+			     " scheduled on another core");
+	}
+}
+
+/**
+ * @brief Verify round robin scheduling on all cores
+ *
+ * @details The test verifies a scenario in which
+ * parent thread joins children to complete task in
+ * round robin.
+ */
+static void test_round_robin_all_cores(void)
+{
```

Comment:
OK, I know you did a little cleanup, but the same complaint applies: this function is EXACTLY THE SAME CODE as test_round_robin_other_cores(), except for:

* It sets the current thread priority first
* It waits for a slightly different timeout (which seems wrong to me -- that shouldn't be a critical number)
* It spawns the threads at a different priority

Please make this a single function that takes two parameters for "spawn_priority" and "current_thread_priority", and make your test_blah_blah_blah() functions one-line wrappers.